Change source, add attributes, and improve state of DirecTV

ehendrix23 opened the pull request:
## Description:

Following enhancements have been made:

1. Added debug logging
2. Added ability to change channel using select_source service of the remote platform.
3. State will now show paused if a recorded program is paused, for live TV playing will always be returned.
4. Added the following attributes:
    a. media_position: current position of the media (in seconds)
    b. media_position_updated_at: timestamp when media_position was updated.
    c. source: current source (channel).
    d. media_isbeingrecorded: if current media is being recorded or not.
    e. media_rating: TV/Movie rating of the media
    f. media_recorded: if current media is recorded or live TV
    g. media_starttime: Timestamp media was aired
5. Reordered properties to follow same order as how they are in __init__.py of remote platform.

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

ehendrix23 committed:
Enhancements for DirecTV media player

Following enhancements have been made:

1. Added debug logging
2. Added ability to change channel using select_source service of the remote platform.
3. State will now show paused if a recorded program is paused, for live TV playing will always be returned.
4. Added the following attributes:
    a. media_position: current position of the media (in seconds)
    b. media_position_updated_at: timestamp when media_position was updated.
   c. source: current source (channel).
   d. media_isbeingrecorded: if current media is being recorded or not.
   e. media_rating: TV/Movie rating of the media
   f. media_recorded: if current media is recorded or live TV
   g. media_starttime: Timestamp media was aired

Reordered properties to follow same order as how they are in __init__.py of remote platform.

ehendrix23 committed:
Fixed error and cleaned up few items

Fixed an issue when determining if a program is recorded or not.
Cleaned up some coding.

MartinHjelmare commented on homeassistant/components/media_player/directv.py:
```diff
@@ -118,13 +162,39 @@ def media_content_id(self):
+        if self._is_standby:
+            return None
+
+        return self._lastupdate
```
This looks like it will spam the state machine even if position hasn't been updated.

ehendrix23 replied:
@MartinHjelmare Thus any updates to attributes (and not just state) will result in the state machine having to do something? 

I also got a different question and I checked the dev docs but could not specifically see it. There are attributes and they are returned by device_state_attributes. But then there are others which are not returned by device_state_attributes (i.e. the media_position_updated_at). What is the difference or best approach for attributes?

Thanks for your insights!

MartinHjelmare replied:
Most properties are returned as part of `state_attributes`, and will create a new state in the state machine when updated. `device_state_attributes` are merged with `state_attributes` in the base entity class before reporting the state to the state machine.

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/media_player/__init__.py#L785-L790

ehendrix23 committed:
Attribute last position update only updated when position changed.

The attribute media_position_updated_at will only be updated if the position changed (thus media is playing for recorded or live TV).
Added assumed_state; will be set to False if in standby or when a recorded show is watched. For live TV it will be set to True.

ehendrix23 committed:
Added some empty lines for easier reading

Added some empty lines before returns to improve readability.

ehendrix23 commented:
@MartinHjelmare Do you mind having a look at this? After my previous commit the travis checked never really finished even although when I clicked on details it said all done and all passed.

Made small update (just added a few empty lines before returns); makes it a bit easier to read as well.
Committed it and now I get the above. But when I check on details, none of the items that I see in there marked red is from direct.py (the one I changed).
What am I missing? :-)

Appreciate it!

MartinHjelmare commented on homeassistant/components/media_player/directv.py:
```diff
@@ -4,26 +4,37 @@
 
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_MEDIA_CURRENTLYRECORDING = 'media_currently_recording'
```
Please separate words with underscore in the constant.

MartinHjelmare commented on homeassistant/components/media_player/directv.py:
```diff
@@ -4,26 +4,37 @@
+ATTR_MEDIA_CURRENTLYRECORDING = 'media_currently_recording'
+ATTR_MEDIA_RATING = 'media_rating'
+ATTR_MEDIA_RECORDED = 'media_recorded'
+ATTR_MEDIA_STARTTIME = 'media_starttime'
```
Borderline case as above.

MartinHjelmare commented on homeassistant/components/media_player/directv.py:
```diff
@@ -88,14 +99,46 @@ def __init__(self, name, host, port, device):
         self._name = name
         self._is_standby = True
         self._current = None
+        self._lastupdate = None
```
See above.

MartinHjelmare replied:
Bump.

MartinHjelmare commented on homeassistant/components/media_player/directv.py:
```diff
@@ -88,14 +99,46 @@ def __init__(self, name, host, port, device):
         self._current = None
+        self._lastupdate = None
+        self._paused = None
+        self._lastposition = None
```
See above.

MartinHjelmare commented on homeassistant/components/media_player/directv.py:
```diff
@@ -88,14 +99,46 @@ def __init__(self, name, host, port, device):
+        self._lastupdate = None
+        self._paused = None
+        self._lastposition = None
+        self._isrecorded = None
```
See above.

MartinHjelmare reviewed (commented):
We can ignore coveralls result. This module is excluded from coverage calculation.

Just tidy the variable names and we can merge.

ehendrix23 committed:
Seperated words in constants

Seperated the words in constants.

ehendrix23 committed:
Fix _lastupdate to _last_update

Split words in _lastupdate to _last_update as I missed it.

ehendrix23 commented:
@MartinHjelmare grmbl, can't believe I missed that one. Updated it. Sorry about that!

MartinHjelmare reviewed (approved):
Good!

MartinHjelmare commented:
Ping me when build is finished.

MartinHjelmare commented:
I'm merging now as the coverage result isn't accurate.
